lib/streamaggr/streamaggr.go

```diff
@@ -774,3 +783,30 @@ func sortAndRemoveDuplicates(a []string) []string {
 	}
 	return dst
 }
+
+// TssUsageTracker tracks used series for streaming aggregation.
+type TssUsageTracker struct {
+	usedSeries map[int]struct{}
+}
+
+// NewTssUsageTracker returns new TssUsageTracker.
+func NewTssUsageTracker(totalSeries int) *TssUsageTracker {
+	return &TssUsageTracker{usedSeries: make(map[int]struct{}, totalSeries)}
+}
+
+// Matched marks series with id as used.
+// Not safe for concurrent use. The caller must
+// ensure that there are no concurrent calls to Matched.
+func (tut *TssUsageTracker) Matched(id int) {
+	tut.usedSeries[id] = struct{}{}
```

Comment:
i'm concerned about number of allocations here... But have no good ideas except of using not `[]prompbmarshal.TimeSeries` but `[]rwTimeSeries` type with additional field `matched`. But I'm not sure it would be more efficient.

Reply:
I addressed this concern in the follow-up commit 52c13e9515652550cd8cc981a7b0331e58e81ff3 . It uses a simple byte slice for tracking matching series, and this byte slice is re-used across calls to streamaggr.Push()